### PR TITLE
[CMake] Add bazel tests for python file lists

### PR DESCRIPTION
**Pull request by Androbin**
Progressing #10296

### CODE REVIEW

**Comment by Androbin** [comment]
@gunan I will write the bazel test. But shouldn't it be triggered by the `cmake` build as well?

**Comment by Androbin** [comment]
I know, the last commit is a bit dirty, but should be enough for some feedback.

**Comment by Androbin** [comment]
The tests should cover the most part. @mrry Any suggestions on further tests?

**Comment by Androbin** [comment]
I let `tf_python.cmake` at least ensure the existence of all entries in the filetree.
As suggested by @gunan I wrote a `BUILD` file to invoke the following:
- `python_modules_test.py` to report invalid and missing/expected entries
- `python_proto_test.py` to report invalid entries
- `python_proto_cc_test.py` to report invalid entries

**Comment by Androbin** [comment]
Also, I fixed the 24 issues reported by `python_modules_test.py`.

**Comment by mrry** [comment]
Thanks for getting this started! I have a few high-level comments about the structure:

* Can you rewrite the Python tests to use `tf.test.TestCase` (similar to other Python `*_test.py` files)? This will make the tests work better with our testing infrastructure, which in turn makes failures easier to track down.
* In their current form,  I think those tests will fail when running on Bazel, because they don't have the appropriate `data` dependencies in the `BUILD` files. FWIW, I expected the test would have dependencies on the appropriate `*.txt` files and a generated PIP package (built using the Bazel target), and it would compare the set of directories. (See [here](https://github.com/tensorflow/tensorflow/blob/3a05c119d8499c3b1e0a3ef6520af3e62738a80c/tensorflow/python/BUILD#L2559) for an example of a test that uses `data` dependencies.)
* Just to check, are the changes in `python_modules.txt` due to code that's been removed?

**Comment by Androbin** [comment]
- I will try to rewrite using `tf.test.TestCase`.
- The tests compare the `.txt` files with the filetree inside `tensorflow/contrib`.
- I understand the concern about requiring the `.txt` files.
- The PIP package shouldn't be needed here?
- Would I need to require `tensorflow/contrib/**` or something?
- Can I acquire a snapshot of the filetree without the file contents?

**Comment by Androbin** [comment]
As for the entries I removed:
- `tensorflow/python/platform/default` -> removed in 2b74220466957e7fbc86ebe3da5ead9a0e4f5169
- `tensorflow/contrib/bayesflow/examples/**` -> removed in 2ccf3aba424405d82e69f03021435e48f54656fb
- `tensorflow/contrib/learn/python/learn/dataframe` -> removed in 27ad708b90cdb62e6aeb7c1378a23ff8cc9c1790
- `tensorflow/contrib/ios_examples/**` -> moved to `tensorflow/examples/ios`

**Comment by Androbin** [comment]
@mrry Can you tell me how to blacklist the dependencies in `pip_smoke_test.py` ?

**Comment by Androbin** [comment]
@mrry Could you test this please?

File: tensorflow/BUILD [hunk 1]
```diff
@@ -395,39 +395,8 @@ py_library(
 )
 
 filegroup(
-    name = "all_opensource_files",
+    name = "all_contrib_files",
```
Comment (gunan) [inline tensorflow/BUILD]: Could you revert this change? we actually have some primitive checks on the all_opensource_files rule, and the flat structure simplifies things.

File: tensorflow/contrib/cmake/BUILD [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+package(default_visibility = ["//visibility:public"])
+
+load("//tensorflow:tensorflow.bzl", "py_test")
+
+py_test(
+    name = "python_modules_test",
+    size = "small",
+    srcs = ["python_modules_test.py"],
+    data = [
+        "//tensorflow:all_contrib_files",
```
Comment (gunan) [inline tensorflow/contrib/cmake/BUILD]: This includes more than what we want in the pip package.
You may want to check this rule instead:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/BUILD#L11

File: tensorflow/contrib/cmake/BUILD [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    ],
```
Comment (gunan) [inline tensorflow/contrib/cmake/BUILD]: You also need to add python_modules.txt as a data dependency here.

File: tensorflow/contrib/cmake/python_modules_test.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+
+from tensorflow.python.platform import test
+
+
+entries_file = "tensorflow/contrib/cmake/python_modules.txt"
```
Comment (gunan) [inline tensorflow/contrib/cmake/python_modules_test.py]: you can use the resource_loader library to access this file.

File: tensorflow/contrib/cmake/BUILD [hunk 5]
```diff
@@ -0,0 +1,46 @@
+licenses(["notice"])  # Apache 2.0
+
+package(default_visibility = ["//visibility:public"])
+
+load("//tensorflow:tensorflow.bzl", "py_test")
+
+py_test(
```
Comment (gunan) [inline tensorflow/contrib/cmake/BUILD]: You also needs the deps field here, and that will need to have the BUILD target for the platform.test library.

**Review by gunan — APPROVED** [review]

**Review by gunan — CHANGES_REQUESTED** [review]
Approved by mistake, please see the comments.

File: tensorflow/BUILD (hunk 1 above)
Comment (Androbin) [reply tensorflow/BUILD]: Is there a more efficient way to require the filetree inside `contrib` ?

File: tensorflow/contrib/cmake/BUILD (hunk 2 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/BUILD]: So, would `//tensorflow/contrib:all_files` work as `deps` ?

File: tensorflow/contrib/cmake/BUILD (hunk 3 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/BUILD]: Had that here, but is already inside `contrib`.

File: tensorflow/contrib/cmake/python_modules_test.py (hunk 4 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/python_modules_test.py]: You mean `self.entries = tf.resource_loader.load_resource(entries_file).split("\n")` ?

File: tensorflow/contrib/cmake/BUILD (hunk 5 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/BUILD]: Should be `//tensorflow/python:platform_test` ?

File: tensorflow/contrib/cmake/BUILD (hunk 3 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/BUILD]: Also, `pip_smoke_test.py` complains about them not being in the PIP package.
Shall I whitelist the `.txt` files or the `_test` targets ?

**Comment by Androbin** [comment]
I have implemented all changes requested by @gunan.
But I still don't know about the `pip_smoke_test.py` whitelisting?

**Comment by Androbin** [comment]
I think I'm not requiring the `contrib` filetree correctly.
It isn't included in the `runfiles` directory.

**Comment by Androbin** [comment]
All files are included correctly now.
But @gunan can't we deduplicate `tensorflow/BUILD` ?

The bazel are only testing entries in `contrib` for now.
This is because `test.TestCase` imports other files (from `runfiles`) which import generated files.

File: tensorflow/contrib/cmake/python_modules_test.py (hunk 4 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/python_modules_test.py]: Trying to remove dependencies to other packages so as not to import from `runfiles` which fails if something references generated files.

**Comment by Androbin** [comment]
This is why I try not to include AND reference other packages:
```
Traceback (most recent call last):
  File "/home/androbin/bazel/tensorflow/bazel-bin/tensorflow/contrib/cmake/python_modules_test.runfiles/org_tensorflow/tensorflow/contrib/cmake/python_modules_test.py", line 22, in <module>
    from tensorflow.python.platform import test
  File "/home/androbin/bazel/tensorflow/bazel-bin/tensorflow/contrib/cmake/python_modules_test.runfiles/org_tensorflow/tensorflow/__init__.py", line 24, in <module>
    from tensorflow.python import *
  File "/home/androbin/bazel/tensorflow/bazel-bin/tensorflow/contrib/cmake/python_modules_test.runfiles/org_tensorflow/tensorflow/python/__init__.py", line 72, in <module>
    from tensorflow.python.ops.standard_ops import *
  File "/home/androbin/bazel/tensorflow/bazel-bin/tensorflow/contrib/cmake/python_modules_test.runfiles/org_tensorflow/tensorflow/python/ops/standard_ops.py", line 64, in <module>
    from tensorflow.python.ops.parsing_ops import *
  File "/home/androbin/bazel/tensorflow/bazel-bin/tensorflow/contrib/cmake/python_modules_test.runfiles/org_tensorflow/tensorflow/python/ops/parsing_ops.py", line 31, in <module>
    from tensorflow.python.ops import gen_parsing_ops
ImportError: cannot import name 'gen_parsing_ops'
```
This happens because Python is referencing the scripts in `runfiles` instead of the PIP package.
@caisq Can you tell how to exclude just enough files to make python not look here?

File: tensorflow/contrib/cmake/python_modules_test.py (hunk 4 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/python_modules_test.py]: Since I need to compute absolute paths elsewhere in the script and `resource_loader` has its quirks, I have adapted the code and placed it directly in the tests.

**Comment by Androbin** [comment]
I have tested this code in various ways. @caisq CI test this please?

File: tensorflow/contrib/cmake/BUILD (hunk 2 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/BUILD]: This dependency isn't included in the PIP package. In fact, `pip_smoke_test.py` complained about it not being included, which is why I had to blacklist this.

**Comment by Androbin** [comment]
@gunan @mrry Any further remarks?

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Comment by Androbin** [comment]
Seems like blacklisting this dependency was not enough:
```
ERROR: /workspace/bazel_pip/tensorflow/contrib/cmake/BUILD:7:1:
	Target '//tensorflow:all_opensource_files' is not visible from
	target '//bazel_pip/tensorflow/contrib/cmake:python_module_test'.
	Check the visibility declaration of the former target
	if you think the dependency is legitimate.
ERROR: /workspace/bazel_pip/tensorflow/contrib/cmake/BUILD:20:1:
	Target '//tensorflow:all_opensource_files' is not visible from
	target '//bazel_pip/tensorflow/contrib/cmake:python_proto_test'.
	Check the visibility declaration of the former target
	if you think the dependency is legitimate.
ERROR: /workspace/bazel_pip/tensorflow/contrib/cmake/BUILD:33:1:
	Target '//tensorflow:all_opensource_files' is not visible from
	target '//bazel_pip/tensorflow/contrib/cmake:python_proto_cc_test'.
	Check the visibility declaration of the former target
	if you think the dependency is legitimate.
```
@gunan You said something about this?
How about modifying [`glob(["**/*.py"])`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/BUILD#L12) to exclude `*_test.py` ?

File: tensorflow/contrib/cmake/BUILD (hunk 5 above)
Comment (gunan) [reply tensorflow/contrib/cmake/BUILD]: Correct.

File: tensorflow/contrib/cmake/python_modules_test.py (hunk 4 above)
Comment (gunan) [reply tensorflow/contrib/cmake/python_modules_test.py]: I am afraid you will need to depend on runfiles, and nothing else for a test that we want to run in an automated way. Because the way we are using bazel is, your test is only guaranteed to have full access to runfiles, and it "won't" have access to the original source tree.
And also, I would recommend going through resource loader. This is because the runfiles location/structure may change. We will maintain resource_loader ourselves to ensure it will keep working the same way it did work before.
If you do not use resource_loader, it is just a 2nd place to maintain this logic. It will be brittle, as runfiles is managed by bazel, and bazel is still not 1.0.

File: tensorflow/contrib/cmake/BUILD (hunk 2 above)
Comment (gunan) [reply tensorflow/contrib/cmake/BUILD]: No, what I meant is, why do you need this here?
Our "golden" source for our pip package files is the bazel build rule. While this "all_opensource_files" rule includes everything, many of these may or may not be in the pip package. While for the simpler case of checking if the cmake entries are valid, this can be useful, checking if the cmake build is in parity with the bazel build is a more interesting check, IMHO.

**Review by gunan — COMMENTED** [review]
The error you mentioned is just a visibility issue. it can be solved by simply modifying the visibility of "all_opensource_files" target in `tensorflow/BUILD` file.
But before that, please see my comments.

File: tensorflow/contrib/cmake/BUILD (hunk 2 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/BUILD]: Comparing CMake to Bazel wasn't my intention. The tests were originally supposed to check validity. Now I am also linting packages that are expected to be included. Analyzing the PIP package was also not planned. All of this can of course be added later.

File: tensorflow/contrib/cmake/python_modules_test.py (hunk 4 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/python_modules_test.py]: The only thing that would be needed is to extract the path resolution into a separate function. This should make everyone happy.

Comment (Androbin) [reply tensorflow/contrib/cmake/python_modules_test.py]: Also, I am aware that only `runfiles` are guaranteed to exist. That is exactly why I introduced the hack I mentioned.

**Comment by Androbin** [comment]
@gunan We can of course fix the visibility and include the tests in the PIP package. But do we even need test scripts in the PIP package? I remember [@mrry saying that we don't want that](https://github.com/tensorflow/tensorflow/pull/14877#issuecomment-347579233).

> Indeed, we probably include too much test code in the current wildcards, and it would be nice to prune it if possible

I think we should just modify [`glob(["**/*.py"])`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/BUILD#L12) to exclude [`*_test.py`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/BUILD#L127) like is done here?

**Comment by Androbin** [comment]
@gunan What I said about runfiles was about same [issue faced when trying to use TensorFlow from within the source directory](https://stackoverflow.com/questions/35953210/error-running-basic-tensorflow-example).
Because the `runfiles` directory has the same issue, I had to introduce [this minor hack](https://github.com/tensorflow/tensorflow/pull/15166/files#diff-627a2a073a8f26eae4124d586f145cc5R30) to make it work.
Not ideal of course, but should be fine. If you wanna suggest a different solution, I keep an open ear.

**Comment by gunan** [comment]
@jart @wicke I am out for a few days. Could you take a look?

**Comment by Androbin** [comment]
Waiting for #15368 and #15315 while redoing bazel dependencies etc.

**Comment by Androbin** [comment]
@jart Maybe we can use a `genquery` instead of requiring `//tensorflow:all_opensource_files` ?

**Comment by Androbin** [comment]
@jart It doesn't seem like we can remove `//tensorflow:all_opensource_files` as dependency from [`tensorflow/contrib/makefile/BUILD`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/makefile/BUILD#L21) and I can't figure out how to have these tests access at least the directory structure either.

**Comment by jart** [comment]
You might be able to do something similar to what our pip package building process does, where we just depend on the things we need, and then we copy everything we want from the foo.runfiles/ symlink tree structure. There are other ways of doing it.

**Comment by Androbin** [comment]
> where we just depend on the things we need

There is no way of knowing what (let's say) [`python_modules.txt`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cmake/python_modules.txt) contains upfront. They are not even guaranteed to be valid which is what these tests are for.

Also, some of the tests depend on knowing about the filetree structure.

> There are other ways of doing it.

Until we find a better alternative for `cmake` (this PR) and `makefile` and all other things that depend on `//tensorflow:all_opensource_files`, I don't see why having it would be an issue.

Even if we were to write a script to dump a summary of the filetree structure into a file and use that, we would still need some way of requiring "all files" for its `runfiles` directory.

**Comment by Androbin** [comment]
This makes we wonder: How does e.g. `tensorflow/contrib/cmake/tf_python.cmake` get access to all files?

**Comment by Androbin** [comment]
@jart What you say?

**Comment by Androbin** [comment]
Waiting for #15315 and #15368 to get merged.

**Comment by Androbin** [comment]
@gunan Are you okay with these changes?
@mrry Do you have further suggestions?

**Comment by Androbin** [comment]
Due to disagreement over #15315, I moved these functions into `test_commons.py` for now.

File: tensorflow/contrib/cmake/python_module_test.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+
+from __init__ import rm_init_py
+rm_init_py(__file__)
+
+import test_commons as commons
```
Comment (gunan) [inline tensorflow/contrib/cmake/python_module_test.py]: (I do not think we need this as a separate file, but)
why not import this as
```
from tensorflow.contrib.cmake import test_commons
```

**Review by gunan — CHANGES_REQUESTED** [review]
I do not need this many python files here.
They can be merged together, and then you won't need to worry about the init file hack, either.

File: tensorflow/contrib/cmake/python_module_test.py (hunk 6 above)
Comment (Androbin) [reply tensorflow/contrib/cmake/python_module_test.py]: > I do not think we need this as a separate file

The tests share a good chunk of their logic, which I de-duplicated to make it usable by potential additional tests.

As for the import:
This is an artifact of experiments I did, due to issues with `__file__`.
But now, I can change that.

**Comment by Androbin** [comment]
> They can be merged together, and then you won't need to worry about the init file hack, either.

The init file hack is about the imports from `tensorflow.python.platform`.

**Comment by Androbin** [comment]
Maybe I need to clarify things about the init file hack:

When TensorFlow is used from within the repository directory, it finds the `__init__.py` files and resolves all imports the the repository directory instead of the PIP package.
The problem is that `tensorflow/__init__.py` imports all submodules including generated ops, which it can't find here.

The same happens when `all_opensource_files` are linked to `runfiles`.

File: tensorflow/contrib/cmake/python_module_test.py (hunk 6 above)
Comment (gunan) [reply tensorflow/contrib/cmake/python_module_test.py]: What I mean is the tests are small enough that we can merge them to exist in a single file.

Comment (Androbin) [reply tensorflow/contrib/cmake/python_module_test.py]: Okay, will do.

**Comment by Androbin** [comment]
@gunan @martinwicke Any updates concerning Sanity Checks?
Do you want to merge this for now and turn it into a sanity check in a follow-up?

**Comment by drpngx** [comment]
@Androbin the sanity check appears to have passed.

**Comment by Androbin** [comment]
@drpngx I was referring to [#15368#issuecomment-353682976](https://github.com/tensorflow/tensorflow/pull/15368#issuecomment-353682976)

**Comment by drpngx** [comment]
I see. @martinwicke @gunan WDYT?

**Comment by jart** [comment]
If you say `data = ["//tensorflow/python", "//tensorflow/contrib:contrib_py"]` then pretty much everything you should need, including the `__init__.py` files, will be available in the runfiles directory.

**Comment by Androbin** [comment]
@jart These tests will become Sanity Check together with most other targets that use `//tensorflow:all_opensource_files`. See https://github.com/tensorflow/tensorflow/pull/15368#issuecomment-354156560 for reference.
This should be the preferred solution.

**Comment by Androbin** [comment]
Closing this in favor of #15670.
